File: leetcode/0001-1000/0901-1000/0981-0990/0988-smallest-string-starting-from-leaf/SmallestStringStartingFromLeaf.java
import java.util.LinkedList;

public class SmallestStringStartingFromLeaf {

  public static void main(String[] args) {
    var solution = new Solution();

    System.out.println(
        solution.smallestFromLeaf(
            new TreeNode(
                0,
                new TreeNode(
                    1,
                    new TreeNode(3),
                    new TreeNode(4)),
                new TreeNode(
                    2,
                    new TreeNode(3),
                    new TreeNode(4)))));

    System.out.println(
        solution.smallestFromLeaf(
            new TreeNode(
                3,
                new TreeNode(9, null, null),
                new TreeNode(20, new TreeNode(15), new TreeNode(7)))));
  }
}

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {
  }

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  public String smallestFromLeaf(TreeNode root) {
    var queue = new LinkedList<Object[]>();
    var smallest = "";

    queue.offer(new Object[] { root, String.valueOf((char) (root.val + 97)) });

    while (!queue.isEmpty()) {
      var objects = queue.poll();
      var node = (TreeNode) objects[0];
      var path = (String) objects[1];

      if (node.left == null && node.right == null) {
        if (smallest.isEmpty() || smallest.compareTo(path) > 0) {
          smallest = path;
        }

        continue;
      }

      if (node.right != null) {
        queue.offer(new Object[] { node.right, (char) (node.right.val + 97) + path });
      }

      if (node.left != null) {
        queue.offer(new Object[] { node.left, (char) (node.left.val + 97) + path });
      }
    }

    return smallest;
  }
}

File: leetcode/0001-1000/0901-1000/0981-0990/0988-smallest-string-starting-from-leaf/SmallestStringStartingFromLeaf.ts
// import {PriorityQueue} from "@data"
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function smallestFromLeaf(root: TreeNode | null): string {
  let smallest = "zzzzzzzzzzzz";

  if (root == null) {
    return "";
  }

  let queue: [TreeNode, string][] = [
    [root, String.fromCharCode(root.val + 97)],
  ];

  while (queue.length > 0) {
    const newQueue: [TreeNode, string][] = [];

    while (queue.length > 0) {
      const [node, accumulated] = queue.shift()!;

      if (node.left == null && node.right == null) {
        smallest =
          smallest.localeCompare(accumulated) > 0 ? accumulated : smallest;
      }

      if (node.left != null) {
        const s = String.fromCharCode(node.left.val + 97) + accumulated;
        newQueue.push([node.left, s]);
      }

      if (node.right != null) {
        const s = String.fromCharCode(node.right.val + 97) + accumulated;
        newQueue.push([node.right, s]);
      }
    }

    queue = newQueue;
  }

  return smallest;
}

console.log(
  smallestFromLeaf(
    new TreeNode(
      2,
      new TreeNode(2, null, new TreeNode(1, new TreeNode(0))),
      new TreeNode(1, new TreeNode(0)),
    ),
  ),
);

// console.log(
//   smallestFromLeaf(
//     new TreeNode(
//       25,
//       new TreeNode(1, new TreeNode(1), new TreeNode(3)),
//       new TreeNode(3, new TreeNode(0), new TreeNode(2)),
//     ),
//   ),
// );
//
// [3,9,20,null,null,15,7]

console.log(
  smallestFromLeaf(
    new TreeNode(
      3,
      new TreeNode(9),
      new TreeNode(20, new TreeNode(15), new TreeNode(7)),
    ),
  ),
);
